src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs

```diff
@@ -0,0 +1,581 @@
+                                // We are already tracking the association of the state machine type to the user code method
+                                // above, so no need to track it here.
+                                Debug.Assert(CompilerGeneratedNames.IsStateMachineType(stateMachineType.Name));
+                                break;
+                            default:
+                                throw new InvalidOperationException();
+                        }
+                    }
+                }
+
+                // Now that we have instantiating methods fully filled out, walk the generated types and fill in the attribute
+                // providers
+                if (_generatedTypeToTypeArgumentInfo != null)
+                {
+                    foreach (var generatedType in _generatedTypeToTypeArgumentInfo.Keys)
+                    {
+                        Debug.Assert(generatedType == generatedType.GetTypeDefinition());
+
+                        if (HasGenericParameters(generatedType))
+                            MapGeneratedTypeTypeParameters(generatedType);
+                    }
+                }
+
+                /// <summary>
+                /// Check if the type itself is generic. The only difference is that
+                /// if the type is a nested type, the generic parameters from its
+                /// parent type don't count.
+                /// </summary>
+                static bool HasGenericParameters(MetadataType typeDef)
+                {
+                    if (typeDef.ContainingType == null)
+                        return typeDef.HasInstantiation;
+
+                    return typeDef.Instantiation.Length > typeDef.ContainingType.Instantiation.Length;
+                }
+
+                void MapGeneratedTypeTypeParameters(MetadataType generatedType)
+                {
+                    Debug.Assert(CompilerGeneratedNames.IsGeneratedType(generatedType.Name));
+                    Debug.Assert(generatedType == generatedType.GetTypeDefinition());
+
+                    if (_generatedTypeToTypeArgumentInfo?.TryGetValue(generatedType, out var typeInfo) != true)
+                    {
+                        // This can happen for static (non-capturing) closure environments, where more than
+                        // nested function can map to the same closure environment. Since the current functionality
+                        // is based on a one-to-one relationship between environments (types) and methods, this is
+                        // not supported.
+                        return;
+                    }
+
+                    if (typeInfo.OriginalAttributes is not null)
+                    {
+                        return;
+                    }
+                    var method = typeInfo.CreatingMethod;
+                    var body = ilProvider.GetMethodIL(method);
+                    if (body is not null)
```

Comment:
Can we end up with a CreatingMethod that doesn't have a body?

Reply:
I think this was added because of PInvokes - but I could be wrong - I'll try to remove it and see what happens.
I agree with you that it would be really weird.

Reply:
I removed it and everything passed... so I guess it can't happen.